crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -687,31 +649,69 @@ impl RecursorDnsHandle {
         &self.record_cache
     }
 
-    async fn append_ips_from_lookup<E: fmt::Display, R: fmt::Debug, I: Iterator<Item = IpAddr>>(
+    async fn append_ips_from_lookup<'a, I: Iterator<Item = &'a NS>>(
         &self,
-        extract_ips: impl Fn(R) -> I,
-        futures: &mut (impl Stream<Item = Result<R, E>> + Unpin),
+        zone: &Name,
+        mut depth: u8,
+        request_time: Instant,
+        nameserver_pool: RecursorPool<TokioRuntimeProvider>,
+        nameservers: I,
         config: &mut NameServerConfigGroup,
-        activity: &str,
-    ) {
+    ) -> Result<u8, Error> {
+        let mut pool_queries = vec![];
+
+        for ns in nameservers {
+            let record_name = ns.0.clone();
+
+            // For child nameservers of zone, we can reuse the pool that was passed in as
+            // nameserver_pool, but for a non-child nameservers we need to get an appropriate pool.
+            let (new_depth, nameserver_pool) = if !crate::is_subzone(zone, &record_name) {
+                self.ns_pool_for_zone(record_name.clone(), request_time, depth)
+                    .await?
+            } else {
+                (depth, nameserver_pool.clone())
+            };
+
+            depth = new_depth;
+
+            pool_queries.push((nameserver_pool, record_name));
+        }
```

Comment:
FWIW, I notice that this PR is switching from only looking up IPs of the first NS record to looking up IPs for all NS records.

I think that the way this is updating `depth` may over-count the depth of the NS recursion done, since each sibling NS record results in adding at least one more to `depth`. Thus, this becomes something in between a count of queries made and a recursive depth. (Both of these would be good to put limits on. RFC 1035 has a recommendation for "a global per-request counter to limit work on a single request", for example.)

Reply:
good catch. It looks like that is down to the call to ns_pool_for_zone in append_ips_from_lookup.  After taking that out, which I believe will be safe, the depth counter doesn't continually increase with additional name server records.

I agree WRT to an overall limit (and do something like that in the draft CNAME branch,) but I think that should be in a separate PR.